components/nginx/scripts/preconfigure.py

```diff
@@ -80,11 +80,54 @@ def _deploy_nginx_config_files():
         )
 
 
-def preconfigure_nginx():
-
+def _deploy_cert_and_key(cert_src, key_src, cert_path, key_path):
+    def _try_deploy(src, dest):
+        try:
+            utils.deploy_blueprint_resource(
+                src_runtime_props[src], dest, NGINX_SERVICE_NAME,
+                user_resource=True, load_ctx=False)
+            return True
+        except Exception as e:
+            if "No such file or directory" in e.stderr:
+                return False
+            else:
+                raise
+    return _try_deploy(cert_src, cert_path), _try_deploy(key_src, key_path)
+
+
+def _deploy_external_cert():
     target_runtime_props = ctx.target.instance.runtime_properties
 
+    external_cert_deployed, external_key_deployed = _deploy_cert_and_key(
+        'rest_certificate', 'rest_key',
+        utils.EXTERNAL_CERT_PATH, utils.EXTERNAL_KEY_PATH)
+
+    if external_key_deployed and external_cert_deployed:
+        ctx.logger.info(
+            'Deployed user-provided external SSL certificate and private key')
+    elif not external_cert_deployed and not external_key_deployed:
+        utils.generate_ssl_certificate(
```

Comment:
Let's maybe take it out to its own function. It's getting a little crowded here.